### ISSUE TITLE
Game can balance a player to the losing team at the very end of the game

### ISSUE THREAD

**Issue by acerspyro** [red-eclipse/base#704]
Since this is very annoying and a big "fuck you" to the player who played for their team, just to get switched to the losing team 9:40 minutes in the game (20 seconds left), I will consider this to be a real bug.

On a side note, players carrying flags can also be suddenly balanced to the other team. This is another big "fuck you", as it means that the player simply drops the flag and disappears.

Team balancing is done in a pretty dumb way right now as it only looks at the stats, thus it can balance anyone (usually the top player of the winning team, which is absolute bs, because that means players like me, Bobbo, Scatter or any of the other skilled players are always going to be the ones to be switched) at very inappropriate moments. The team balancing functionality needs more brains.

Can be seen happening in current release, all the way to master.

**Comment by TheAssassin** [comment]
First of all, since this is a feature request, you should've posted it on the forums.

The problem with team balancing is that it is really hard to come up with a better approach. Everything has advantages and disadvantages. We've discussed this topic a lot in-game. We couldn't reach a consent a single time.

Most suggestions involved protecting players from a switch after they scored or otherwise deprioritize the ones who've had most positive influence on their team's score. Unfortunately, that could lead to a situation in which better players are never going to be switched. Also, e.g. switching less good players to the losing team isn't really balancing anything.

Suggestions welcome.

**Comment by TheLastProject** [comment]
Why is this a "fuck you"? Is the issue really the balancing, or is it users not getting enough acknowledgement of their achievements? I feel just showing badges on players on the scoreboard like "most captures" seems like a much better way to acknowledge players for their achievements and make them feel rewarded?

**Comment by TheAssassin** [comment]
This whole situation started to become a problem when the stats were introduced. Players on a server are able to see a player's achievements, whatever team the player was in.

The way the stats are currently implemented doesn't depict the situation very well. They also discriminate players in situations @acerspyro described. I am not such a big fan of that system anyway, it's got lots of issues regarding privacy, legal issues as a registered player cannot turn them off, they're way too slow, etc etc etc. I have some ideas how to fix these things, but as long as we're not running things on our own hardware, it's almost impossible to fix these things. It's time for us to move to our own server, @qreeves!

Nevertheless, the balancing itself could work better in my opinion, too. I just can't come up with proper solution. That's why I'm waiting for @acerspyro to post ideas.

**Comment by Robalni** [comment]
I don't think we can make this perfect but there are some things we can do to make it less annoying. One thing is to not balance the teams after half time because that doesn't make much difference anyway other than annoying players, and it should not switch players when they carry a flag or bomb or secure control points.

**Comment by acerspyro** [comment]
I like @TheLastProject 's ideas, and @Robalni 's ideas should definitely be implemented

**Comment by shacknetisp** [comment]
> Team balancing is done in a pretty dumb way right now as it only looks at the stats, thus it can balance anyone (usually the top player of the winning team, which is absolute bs, because that means players like me, Bobbo, Scatter or any of the other skilled players are always going to be the ones to be switched) at very inappropriate moments. 

This is not entirely true. Team balancing is roughly 50% stats, 50% current performance. The best players are moved first because the variable `teambalancehighest` is set to 1, IIRC before stats were ever introduced and certainly before they were used for balance.

> This whole situation started to become a problem when the stats were introduced.

Balance has always been horrible, but before 1.4.5 or so the system was essentially random, resulting in *better* gameplay than with these sorted teams. Statistics have very little to do with this, they only allow for a longer balancing average than just the server's history of a player. (Random balancing again, anyone?)

> Players on a server are able to see a player's achievements, whatever team the player was in.

I'm not sure why this is a problem, or a new problem. People have always known who are good players and who are mediocre players, the ability to turn on statistics information displays hasn't changed that.

> I don't think we can make this perfect but there are some things we can do to make it less annoying. One thing is to not balance the teams after half time because that doesn't make much difference anyway other than annoying players, and it should not switch players when they carry a flag or bomb or secure control points.

Here are some useful ideas, even disregarding the algorithm these could improve play.

The statistics system should probably be discussed elsewhere.

**Comment by bonifarz** [comment]
I agree with @Robalni's suggestions, too.

Imho, auto-balance is a useful feature, its implementation is certainly not "dumb", but maybe some tweaks will help to improve things.

**Comment by shacknetisp** [comment]
I've thought of some ways to improve team balance. Currently when there's an imbalance in players the team with the most players will always lose a player to the team with least players (by default, this will be the highest score player that will be switched.) Furthermore, due to the way players joining get their teams chosen, the worst team will always have the new player join them, meaning the worst team will have their best player switched.

This can be solved by moving more than one player when required, specifically three. The server should, instead of moving the highest/lowest balscore player, move through all possibilities of moving one player from the team with more players OR two players from the team with more players and one from the team with less players. It then performs the action that will result in team total balscores moving closer together, thereby *actually* balancing the game. (Multi adds two more teams, I'm not entirely sure how to apply this there, but the same basic principles are the same.)

Now this will cause some disruptions by moving more players, so restrictions and priorities must be implemented. When moving three players, for instance, one of those three should be the recently-joined player. Players who have contributed a lot to their team or have played for a while in the game should be de-prioritized. There are other possibilities, but these are only conveniences to avoid too much disruption.

I consider the team balancing algorithm to be the worst flaw in Red Eclipse, it's hard to enjoy team games when they are <strike>always</strike>often one-sided massacres when players begin joining after the initial team selection. Fixing this should be extremely prioritized before the next release.
